src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs

```diff
@@ -133,58 +134,96 @@ private Task<ChecksumCollection> BuildAsync(IReadOnlyList<AnalyzerReference> key
             return CreateChecksumCollectionsAsync(documentStates, kind, snapshotBuilder.BuildAsync, cancellationToken);
         }
 
-        private Task<ChecksumCollection> CreateChecksumCollectionsAsync(
+        private async Task<ChecksumCollection> CreateChecksumCollectionsAsync<TValue, TChecksumObject>(
+            IEnumerable<TValue> items, string kind, Func<TValue, CancellationToken, Task<TChecksumObject>> buildAsync, CancellationToken cancellationToken) where TChecksumObject : ChecksumObject
+        {
+            using (var pooledObject = Creator.CreateList<Task<TChecksumObject>>())
+            {
+                // create asyn checksums concurrently
+                var tasks = pooledObject.Object;
+                foreach (var item in items)
+                {
+                    tasks.Add(buildAsync(item, cancellationToken));
+                }
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+
+                var checksums = new Checksum[tasks.Count];
+
+                for (var i = 0; i < tasks.Count; i++)
+                {
+                    var task = tasks[i];
+
+                    // task is already completed. so wait will return right away as long as there is no exception.
+                    // if there is exception, then it will throw it. if it is cancellation exception, it will make sure
+                    // it throws cancellation exception rather than aggregated exception.
+                    task.Wait(cancellationToken);
```

Comment:
We can address confusion with a comment.  The problem with teh current approach is that it now means we have to deal with this block of code every time we do an async/sync audit.  IDE does not want blocking calls withing async blocks.  It violates rules we have about how we do async.  Even if this is 'ok' it still adds a tax as we have to now consider this code each and every time and give it a pass.

There is no reason for this code to be special and to not follow the rules we have about async.  @AdamSpeight2008's confusion can be resolved with comments if we feel that's necessary.  But the code should be using 'awaits' to fit in with how we want all our code to be at these layers.